### PR TITLE
Fix optimizer parameter buffer allocation size.

### PR DESCRIPTION
**Pull request by helinwang**
The buffer allocation size should be number of bytes, not number of
floats.

Thanks @reyoung for noticing the bug!

Fixes: https://github.com/PaddlePaddle/Paddle/issues/2854
Fixes: #2858

### CODE REVIEW

File: go/pserver/optimizer.go [hunk 1]
```diff
@@ -78,7 +80,11 @@ func (o *optimizer) UpdateParameter(g Gradient) error {
 		return fmt.Errorf("Name: %s, parameter and gradient element type not match, parameter: %v, gradient: %v", g.Name, o.elementType, g.ElementType)
 	}
 
-	r := C.paddle_update_parameter(o.opt, C.paddle_element_type(g.ElementType), unsafe.Pointer(&g.Content[0]), C.int(len(g.Content))/C.sizeof_float)
+	if o.contentLen != len(g.Content) {
+		return fmt.Errorf("Name: %s, parameter and gradient does not have same content len, parameter: %d, gradient: %d", g.Name, o.contentLen, len(g.Content))
+	}
+
+	r := C.paddle_update_parameter(o.opt, C.paddle_element_type(g.ElementType), unsafe.Pointer(&g.Content[0]), C.int(len(g.Content)))
```
Comment (typhoonzero) [inline go/pserver/optimizer.go]: In `optimizer/optimizer.cc`, Content is casted to float array, and length(num_bytes) is passed to `Tensor.h`
```c++
int paddle_update_parameter(paddle_optimizer* o,
                            const paddle_element_type data_type,
                            const void* grad_buffer,
                            int num_bytes) {
  auto grad_type = reinterpret_cast<const float*>(grad_buffer);
  Tensor* gradient = new Tensor(const_cast<float*>(grad_type), num_bytes);
```
in `optimizer/Tensor.h`:

```c++
TensorT(T* data, size_t size)
      : height_(1), width_(size), data_ptr_(nullptr), data_(data) {}
```

The num_bytes is used as width, which means count of float numbers. So this is not bytes.

Maybe we should change the parameter name of `paddle_update_parameter` `num_bytes` to `array_size`

File: go/pserver/optimizer.go [hunk 2]
```diff
@@ -37,10 +38,11 @@ func cArrayToSlice(p unsafe.Pointer, len int) []byte {
 func newOptimizer(paramWithConfigs ParameterWithConfig, State []byte) *optimizer {
 	o := &optimizer{}
 	o.elementType = paramWithConfigs.Param.ElementType
+	o.contentLen = len(paramWithConfigs.Param.Content)
 	p := paramWithConfigs.Param
 	c := paramWithConfigs.Config
 	s := State
-	paramBufferSize := C.size_t(len(p.Content) / C.sizeof_float)
+	paramBufferSize := C.size_t(len(p.Content))
```
Comment (typhoonzero) [inline go/pserver/optimizer.go]: Well, malloc size is indeed must be the length of `p.Content` the type is `unsigned char*` means a byte buffer.

File: go/pserver/optimizer.go (hunk 1 above)
Comment (dzhwinter) [inline go/pserver/optimizer.go]: @typhoonzero helin move the size changing to tensor side.  may be it is more clear put this near type casting.

**Comment by typhoonzero** [comment]
Related: https://github.com/PaddlePaddle/Paddle/pull/2859

File: go/pserver/optimizer.go (hunk 1 above)
Comment (helinwang) [inline go/pserver/optimizer.go]: `num_bytes` is for `const void* grad_buffer` which means the length of the buffer. Then `paddle_update_parameter` need to calculate the array_size based of the type of the element. That's why I changed from `new TensorT(buffer, num_bytes)` to `new Tensor(buffer, num_bytes/sizeof(float))` [here](https://github.com/PaddlePaddle/Paddle/pull/2855/files#diff-defc4d95efae04a339f807e16eb7265aR47)

File: go/pserver/client/client_test.go [hunk 3]
```diff
@@ -100,66 +102,110 @@ func (l lister) List() []client.Server {
...
+		j := rand.Intn(i + 1)
+		grads[i], grads[j] = grads[j], grads[i]
+	}
+
+	for i := 0; i < numGroups; i++ {
+		var gs []pserver.Gradient
+		if i == numGroups-1 {
+			gs = grads[i*paramPerGroup:]
+		} else {
+			gs = grads[i*paramPerGroup : (i+1)*paramPerGroup]
```
Comment (dzhwinter) [inline go/pserver/client/client_test.go]: just curious, if we send grads with the size of 0, 10000, 0, 10000 like a sawtooth. compare with group similar size together, like 0, 0 ,0, 10000, 10000. Is it hurt the performance?

File: go/pserver/optimizer.go (hunk 1 above)
Comment (typhoonzero) [inline go/pserver/optimizer.go]: Sorry, I didn't see the change below.

File: go/pserver/client/client_test.go [hunk 4]
```diff
@@ -100,66 +102,110 @@ func (l lister) List() []client.Server {
...
-		if err != nil {
-			t.Fatal(err)
-		}
+		wg.Add(1)
+		go func(i int) {
+			var p pserver.Parameter
+			p.Name = "p_" + strconv.Itoa(i)
+			p.ElementType = pserver.Float32
+			p.Content = make([]byte, (i+1)*100)
+			err := c.InitParam(pserver.ParameterWithConfig{Param: p, Config: config})
```
Comment (typhoonzero) [inline go/pserver/client/client_test.go]: Do we need multiple client instances to simulate multiple client to init parameters at the same time?

Comment (helinwang) [inline go/pserver/client/client_test.go]: Great idea! I will send a follow up PR for this one.

File: go/pserver/client/client_test.go (hunk 3 above)
Comment (helinwang) [inline go/pserver/client/client_test.go]: Great question, This is for testing concurrent send. I think the system performs best when the traffic for each parameter server is similar size.
the current implementation will send one entire parameter  (no matter how big it is) to one parameter server. In the near future will will make the optimization to partition each big parameter into smaller chunks and shard them to different parameter servers. In this way, the traffic for each parameter server is similar size.

**Pull request merged**